### PR TITLE
Add option to use FILE * streams with cf_popen_full_duplex

### PR DESCRIPTION
**Pull request by olehermanse**
Merge together:
https://github.com/cfengine/core/pull/4180
https://github.com/cfengine/enterprise/pull/616

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5137/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5137/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5137/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5138/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5138/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5138/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5139/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5139/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5139/

File: libpromises/pipes_unix.c [hunk 1]
```diff
@@ -879,17 +881,32 @@ int cf_pclose_full_duplex_side(int fd)
...
 
-        if (data->write_fd >= 0)
+        if (data->write_stream != NULL)
+        {
+            // fclose closes the underlying fd / socket,
+            // but we need the fd for handling of processes below.
+            assert(data->write_fd >= 0);
+            fclose(data->write_stream);
+        }
+        else if (data->write_fd >= 0)
```
Comment (vpodzime) [inline libpromises/pipes_unix.c:909]: Should this maybe set `data->write_stream` and `data->write_fd` to `NULL` and `0` respectively?

Comment (olehermanse) [reply libpromises/pipes_unix.c:909]: It could, but it didn't before, so I didn't change it. (Also, if changed, it should be `-1` for invalid fd.)

**Pull request merged**